### ISSUE TITLE
Decorators cause unwanted warnings about = vs == when code is run in gjs

### ISSUE THREAD

**Issue by realh** [microsoft/TypeScript#31774]
**TypeScript Version:**  3.6.0-dev.20190604

**Search Terms:**
test for equality (==) mistyped as assignment (=)?
test for equality (==) mistyped as assignment (=)? decorator

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
// Requires --experimentalDecorators=true

function deco(klass: Function) {
}

@deco
class TestClass {
}
```

**Expected behavior:**
When the transpiled script is run it should not print any warnings.

**Actual behavior:**
When run in gjs (which is based on Mozilla Spidermonkey) it causes a warning like this:
`Gjs-Message: 13:59:43.346: JS WARNING: [index.js 4]: test for equality (==) mistyped as assignment (=)?`

It's because the generated code includes this if clause: `if (d = decorators[i])`. I don't think this is an error, but I think it would be better to write it as `if ((d = decorators[i]) != null)` to make its intention clear and prevent false alarms which end users won't want to see.

**Playground Link:** 

**Related Issues:** 
No, but it would probably be a good idea to test at least other types of decorator to check that they don't independently cause the same issue.

**Comment by fatcerberus** [comment]
Hmm, while I see your point, the question with this kind of thing is, where do you draw the line?  What you’re essentially seeing is a lint warning, and TypeScript can’t possibly satisfy every linter and lint configuration in the world.  At the end of the day, tsc is a compiler; as long as the code it produces isn’t misbehaving at runtime, you shouldn’t need to care what the generated code looks like.

That being said, it’s unfortunate that in this particular case, some kind of linting *is* being done at runtime, which IMO is pretty bizarre, especially in this day and age of transpilers and minification everywhere.

**Comment by nattthebear** [comment]
Linting generated code seems doomed to failure.  But, the TS team has accepted such changes before:  https://github.com/microsoft/TypeScript/issues/27312

**Comment by fatcerberus** [comment]
@natthebear Yeah, the thing here is, there’s some runtime (this “gjs”) which automatically lints the code it runs - so it’s a different case than a developer *choosing* to lint the generated code.  The user actually has no choice in the matter (which is misguided on gjs part but what can you do).

Alas, the complexities inherent in JS being both “the assembly language of the web” and “source code” simultaneously...

**Comment by realh** [comment]
gjs provides interoperability between SpiderMonkey and gobject-introspection, so you can write applications in JS using libraries like GTK and GStreamer. The warnings come from SpiderMonkey, so  I suspect they would also appear in Firefox's developer console if TS decorators were used in a web app, but I haven't verified that.

I turns out [gjs does have a way I can disable these warnings](https://gitlab.gnome.org/GNOME/gjs/issues/254) so I don't mind personally if you want to close this, but I'll leave the final decision up to you.

**Comment by RyanCavanaugh** [comment]
The code is correctly written, and is written in a way to be as short as possible to minimize download size. A runtime that errors on a legal assignment isn't a JS runtime and we're not capable of forming our output to conform to *all* third-party opinions.
